source/Irrlicht/CQuantNormalCache.cpp

```diff
@@ -0,0 +1,181 @@
+
+		return restoreSign<uint32_t>(vec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	uint32_t absBestFit = absIntFit[0] | (absIntFit[1] << quantizationBits) | (absIntFit[2] << (quantizationBits * 2u));
+
+	normalCacheFor2_10_10_10Quant.insert(std::make_pair(mapToBarycentric(core::abs(normal)), absBestFit));
+
+	return restoreSign<uint32_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
+
+uint32_t CQuantNormalCache::quantizeNormal8_8_8(const core::vectorSIMDf& normal)
+{
+	constexpr uint32_t quantizationBits = 8u;
+	const auto xorflag = core::vectorSIMDu32((0x1u << quantizationBits) - 1u);
+	auto negativeMask = normal < core::vectorSIMDf(0.f);
+
+	auto found = normalCacheFor8_8_8Quant.find(mapToBarycentric(normal));
+	if (found != normalCacheFor8_8_8Quant.end() && (found->first == mapToBarycentric(normal)))
+	{
+		const auto absVec = core::vectorSIMDu32(found->second.x, found->second.y, found->second.z);
+
+		return restoreSign<uint32_t>(absVec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	Vector8u bestFit = { absIntFit[0], absIntFit[1], absIntFit[2] };
+
+	normalCacheFor8_8_8Quant.insert(std::make_pair(mapToBarycentric(normal), bestFit));
+
+	return restoreSign<uint32_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
+
+uint64_t CQuantNormalCache::quantizeNormal16_16_16(const core::vectorSIMDf& normal)
+{
+	constexpr uint32_t quantizationBits = 10u;
+	const auto xorflag = core::vectorSIMDu32((0x1u << quantizationBits) - 1u);
+	auto negativeMask = normal < core::vectorSIMDf(0.f);
+
+	auto found = normalCacheFor16_16_16Quant.find(mapToBarycentric(normal));
+	if (found != normalCacheFor16_16_16Quant.end() && (found->first == mapToBarycentric(normal)))
+	{
+		const auto absVec = core::vectorSIMDu32(found->second.x, found->second.y, found->second.z);
+
+		return restoreSign<uint64_t>(absVec, xorflag, negativeMask, quantizationBits);
+	}
+
+	core::vectorSIMDf fit = findBestFit(quantizationBits, normal);
+	auto absIntFit = core::vectorSIMDu32(core::abs(fit)) & xorflag;
+
+	Vector16u bestFit = { absIntFit[0], absIntFit[1], absIntFit[2] };
+
+	normalCacheFor16_16_16Quant.insert(std::make_pair(mapToBarycentric(normal), bestFit));
+
+	return restoreSign<uint64_t>(absIntFit, xorflag, negativeMask, quantizationBits);
+}
+
+void CQuantNormalCache::insertIntoCache2_10_10_10(const core::vectorSIMDf& key, const uint32_t quantizedNormal)
+{
+
+}
+void CQuantNormalCache::insertIntoCache8_8_8(const core::vectorSIMDf& key, const uint32_t quantizedNormal)
+{
+
+}
+void CQuantNormalCache::insertIntoCache16_16_16(const core::vectorSIMDf& key, const uint32_t quantizedNormal)
+{
+
+}
```

Comment:
you could template these functions and turn them into one, by templating on the cache type (enum or stored value type)

Reply:
not anymore

Reply:
not after recent changes